packages/web/src/common/store/premiumContent/sagas.ts

```diff
@@ -329,14 +336,102 @@ function* pollPremiumTrack({
   while (true) {
     const premiumTrackSignatureMap = yield* select(getPremiumTrackSignatureMap)
     if (premiumTrackSignatureMap[trackId]) {
-      yield* put(updatePremiumTrackStatus({ status: 'UNLOCKED' }))
+      yield* put(updatePremiumTrackStatus({ trackId, status: 'UNLOCKED' }))
       break
     }
     yield* put(trackPageActions.fetchTrack(null, slug, handle, false, true))
     yield* delay(frequency)
   }
 }
 
+/**
+ * 1. Get follow or tip gated tracks of user
+ * 2. Set those track statuses to 'UNLOCKING'
+ * 3. Poll for premium content signatures for those tracks
+ * 4. When the signatures are returned, set those track statuses as 'UNLOCKED'
+ */
+function* updateGatedTracks(trackOwnerId: ID, gate: 'follow' | 'tip') {
+  const statusMap: { [id: ID]: PremiumTrackStatus } = {}
+  const trackParamsMap: { [id: ID]: TrackRouteParams } = {}
+  const cachedTracks = yield* select(getTracks, {})
+
+  Object.keys(cachedTracks).forEach((trackId) => {
+    const id = parseInt(trackId)
+    const {
+      owner_id: ownerId,
+      permalink,
+      premium_conditions: premiumConditions
+    } = cachedTracks[id]
+    const isGated =
+      gate === 'follow'
+        ? premiumConditions?.follow_user_id
+        : premiumConditions?.tip_user_id
+    if (isGated && ownerId === trackOwnerId) {
+      statusMap[id] = 'UNLOCKING'
+      trackParamsMap[id] = parseTrackRoute(permalink)
+    }
+  })
+
+  yield* put(updatePremiumTrackStatuses(statusMap))
+
+  yield* all(
+    Object.keys(trackParamsMap).map((trackId) => {
+      const id = parseInt(trackId)
+      return call(pollPremiumTrack, {
+        trackId: id,
+        trackParams: trackParamsMap[id],
+        frequency: PREMIUM_TRACK_POLL_FREQUENCY
+      })
+    })
+  )
+}
+
+/**
+ * 1. Get follow-gated tracks of unfollowed user
+ * 2. Set those track statuses to 'LOCKED'
+ * 3. Remove the premium content signatures for those tracks
+ */
+function* handleUnfollowUser(
+  action: ReturnType<typeof usersSocialActions.unfollowUser>
+) {
+  const statusMap: { [id: ID]: PremiumTrackStatus } = {}
+  const cachedTracks = yield* select(getTracks, {})
+
+  Object.keys(cachedTracks).forEach((trackId) => {
+    const id = parseInt(trackId)
+    const { owner_id: ownerId, premium_conditions: premiumConditions } =
+      cachedTracks[id]
+    const isFollowGated = premiumConditions?.follow_user_id
+    if (isFollowGated && ownerId === action.userId) {
+      statusMap[id] = 'LOCKED'
+    }
+  })
+
+  yield* put(updatePremiumTrackStatuses(statusMap))
+
+  const trackIds: ID[] = []
+  const metadatas: { id: ID; metadata: Partial<TrackMetadata> }[] = []
+  Object.keys(statusMap).forEach((trackId) => {
+    const id = parseInt(trackId)
+    trackIds.push(id)
+    metadatas.push({ id, metadata: { premium_content_signature: null } })
+  })
+  yield* put(removePremiumContentSignatures({ trackIds }))
```

Comment:
Feels like it would be easy at some point to introduce a bug by doing one of these actions without the other - I wonder if `removePremiumContentSignature` should have the side-effect of making sure that the cached metadata _also_ removes the premium content signature? Wdyt?

Reply:
agreed and handled